File: statistics.ml
(* statistics.ml *)
(* This OCaml script calculates mean, median, and mode using a functional paradigm.
   It uses immutable lists, pattern matching, and higher-order functions. *)

(* Function to calculate mean *)
let mean lst =
  let total = List.fold_left (+.) 0.0 (List.map float_of_int lst) in
  total /. float_of_int (List.length lst)

(* Function to calculate median *)
let median lst =
  let sorted = List.sort compare lst in
  let n = List.length sorted in
  if n mod 2 = 0 then
    let mid1 = List.nth sorted (n / 2 - 1)
    and mid2 = List.nth sorted (n / 2) in
    float_of_int (mid1 + mid2) /. 2.0
  else
    float_of_int (List.nth sorted (n / 2))

(* Function to calculate mode(s) *)
let mode lst =
  let freq = List.fold_left (fun acc x ->
    let count = try List.assoc x acc + 1 with Not_found -> 1 in
    (x, count) :: List.remove_assoc x acc
  ) [] lst in
  let max_freq = List.fold_left (fun acc (_, c) -> max acc c) 0 freq in
  List.fold_left (fun acc (x, c) -> if c = max_freq then x::acc else acc) [] freq

(* Main entry point with test data *)
let () =
  let data = [1; 3; 2; 4; 2; 5; 2] in
  Printf.printf "Mean: %.2f\n" (mean data);
  Printf.printf "Median: %.2f\n" (median data);
  Printf.printf "Mode(s): ";
  List.iter (Printf.printf "%d ") (mode data);
  Printf.printf "\n"

File: statistics.c
// statistics.c
// This C program calculates mean, median, and mode from a list of integers.
// Implements a procedural approach using arrays and standard C functions.

#include <stdio.h>
#include <stdlib.h>

// Comparison function for qsort
int compare(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

// Function to calculate mean
float mean(int arr[], int n) {
    int sum = 0;
    for(int i = 0; i < n; i++) sum += arr[i];
    return (float)sum / n;
}

// Function to calculate median
float median(int arr[], int n) {
    qsort(arr, n, sizeof(int), compare); // Sort the array
    if(n % 2 == 0)
        return (arr[n/2 - 1] + arr[n/2]) / 2.0;
    else
        return arr[n/2];
}

// Function to calculate and print mode
void mode(int arr[], int n) {
    int maxCount = 0;
    printf("Mode(s): ");
    for(int i = 0; i < n; i++) {
        int count = 1;
        for(int j = i+1; j < n; j++) {
            if(arr[j] == arr[i]) count++;
        }
        if(count > maxCount) maxCount = count;
    }

    // Print all modes with max frequency
    for(int i = 0; i < n; i++) {
        int count = 1;
        for(int j = i+1; j < n; j++) {
            if(arr[j] == arr[i]) count++;
        }
        if(count == maxCount) {
            printf("%d ", arr[i]);
        }
    }
    printf("\n");
}

int main() {
    int arr[] = {1, 3, 2, 4, 2, 5, 2};  // Example dataset
    int n = sizeof(arr) / sizeof(arr[0]);

    printf("Mean: %.2f\n", mean(arr, n));
    printf("Median: %.2f\n", median(arr, n));
    mode(arr, n);

    return 0;
}
